hibernate-core/src/test/java/org/hibernate/orm/test/annotations/generics/EmbeddedIdGenericsSuperclassTest.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.orm.test.annotations.generics;
+
+import java.io.Serializable;
+import java.util.Random;
+
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.Jira;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.EmbeddedId;
+import jakarta.persistence.Entity;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Marco Belladelli
+ */
+@DomainModel(annotatedClasses = {
+		EmbeddedIdGenericsSuperclassTest.Customer.class,
+		EmbeddedIdGenericsSuperclassTest.Invoice.class,
+		EmbeddedIdGenericsSuperclassTest.NestedCustomer.class
+})
+@SessionFactory
+@Jira("https://hibernate.atlassian.net/browse/HHH-16188")
+public class EmbeddedIdGenericsSuperclassTest {
+	@BeforeAll
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final Customer customer = new Customer( 1, "Francisco" );
+			session.persist( customer );
+			final Invoice invoice = new Invoice( 2 );
+			session.persist( invoice );
+			final NestedCustomer nestedCustomer = new NestedCustomer( "Marco" );
+			session.persist( nestedCustomer );
+		} );
+	}
+
+	@AfterAll
+	public void tearDown(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			session.createMutationQuery( "delete from Customer" ).executeUpdate();
+			session.createMutationQuery( "delete from Invoice" ).executeUpdate();
+			session.createMutationQuery( "delete from NestedCustomer" ).executeUpdate();
+		} );
+	}
+
+	@Test
+	public void testCustomer(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final Customer customer = session.createQuery(
+					"from Customer c where c.id.someDomainField = 1",
+					Customer.class
+			).getSingleResult();
+			assertThat( customer ).isNotNull();
+			assertThat( customer.getCode() ).isEqualTo( 1 );
+			assertThat( customer.getName() ).isEqualTo( "Francisco" );
+		} );
+	}
+
+	@Test
+	public void testCustomerCriteria(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final CriteriaBuilder cb = session.getCriteriaBuilder();
+			final CriteriaQuery<Customer> query = cb.createQuery( Customer.class );
+			final Root<Customer> root = query.from( Customer.class );
+			query.select( root ).where( cb.equal( root.get( "id" ).get( "someDomainField" ), 1 ) );
```

Comment:
Please add some assertions on the type of the `root.get("id")` path to ensure we don't leak the generic type somehow.

Reply:
Done (thanks to new logic, see other comment).